Custom paging

guy120494 opened the pull request:
As discussed in #160 , this PR allows to specify each connection the type of the connection.
It also has a support for simple paging (without edges and nodes)

@yarinvak @tdraier please code review me.

guy120494 committed:
1. Deleted the license plugin
2. Added hamcrest

guy120494 committed:
Refactor

guy120494 committed:
Added custom relay for simple paging

guy120494 committed:
Rename

guy120494 committed:
Added the option to specify custom relay

guy120494 committed:
Moved to its own package

guy120494 committed:
Added simple connection support

guy120494 committed:
Added classes

guy120494 committed:
Added this class to be as a default class in the GraphQLConnection annotation. This class does nothing, If you don't specify a custom connection, the connection in the container is used

guy120494 committed:
Use the right relay

guy120494 committed:
Added a "getRelay" method

guy120494 committed:
Change doc

guy120494 committed:
More support to SimplePaging

guy120494 committed:
Removed unnecessary method

guy120494 committed:
Clear the container and the type registry after done

guy120494 committed:
Moved test

guy120494 committed:
Simple connection extends connection because we need to return "Connection"
object (The irrelevant methods throw appropriate exception)

guy120494 committed:
Data type is wrapped with GraphQLList

guy120494 committed:
Merge branch 'master' into custom-paging

guy120494 committed:
Fix test

guy120494 committed:
Update readme

guy120494 committed:
Fix test

guy120494 committed:
Clear container and type registry only if exception is thrown

guy120494 committed:
Fix

guy120494 committed:
Added java doc, and made it final

tdraier commented on src/main/java/graphql/annotations/connection/GraphQLConnection.java:
```diff
@@ -59,4 +61,12 @@
+     *
+     * @return a class that represents the connection type
+     */
+    Class<? extends Relay> connectionType() default FakeRelay.class;
```
Why do you need a default and FakeRelay here ? FakeRelay is never instantiated, you could just keep an undefined connectionType and test for null in ConnectionUtil

guy120494 replied:
I want the default connectionType to be taken from `GraphQLAnnotations` class, so I have to define a fake default so I will know when to take the default.
When I don't specify a default type, every time I use the `GraphQLConnection` annotation I must specify a `connectionType`, and I don't want to do that

tdraier replied:
Ok, I thought it was possible to have no default value and don't make it mandatory, my mistake

tdraier commented on src/main/java/graphql/annotations/connection/GraphQLConnection.java:
```diff
@@ -37,7 +39,7 @@
      * @return a connection class
      */
-    Class<? extends ConnectionFetcher> connection() default PaginatedDataConnectionFetcher.class;
+    Class<? extends ConnectionFetcher> connectionFetcher() default PaginatedDataConnectionFetcher.class;
```
I don't know if it's really an issue, but it breaks compatibility and should be marked deprecated first. But I agree connectionFetcher is more accurate.

guy120494 replied:
Well, We don't upgrade this library often, so I think it is OK

tdraier replied:
Ok for me too

tdraier commented on src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java:
```diff
@@ -1,17 +1,3 @@
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- */
```
What happened to the license header ?

guy120494 replied:
I don't see a point to mention the license anymore, because the founder of the library is not active anymore, and the library have been changed a lot since.
But I don't familiar with the rules, so if it's not legal/legitimate, I will put it back

tdraier commented on build.gradle:
```diff
@@ -11,7 +11,6 @@ buildscript {
     }
 }
 plugins {
-    id "com.github.hierynomus.license" version "0.12.1"
```
We don't check license anymore ?

guy120494 replied:
I don't see a point to mention the license anymore, because the founder of the library is not active anymore, and the library have been changed a lot since.
But I don't familiar with the rules, so if it's not legal/legitimate, I will put it back

tdraier replied:
I have no idea neither about the license, but I prefer to keep the Apache license header everywhere and keep the plugin to enforce it

guy120494 replied:
Put it back

tdraier commented on src/main/java/graphql/annotations/connection/simple/SimpleConnectionFetcher.java:
```diff
@@ -0,0 +1,6 @@
+import graphql.annotations.connection.ConnectionFetcher;
+
+public interface SimpleConnectionFetcher<T> extends ConnectionFetcher<T> {
+}
```
What's the point of this interface ? I don't see any usage which could not be replaced by ConnectionFetcher directly

guy120494 replied:
It's just for the name, so I know what interface to extend when Im implementing the simple connection, but I can delete this interface

tdraier commented on src/main/java/graphql/annotations/connection/simple/SimpleConnection.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * This is a more intuitive implementation for paging. You send back only the data that is requested,
+ * along with "overall" - the amount of all the entities
+ *
```
It does not look "more" simple since it extends Connection, and that you still need to implement getEdges() and getPageInfo() .. ? Edges and PageInfo are also not needed by SimpleRelay.

guy120494 replied:
It's simple to use. The problem is that to use the exists annotation for simple paging, I needed to extend `Relay`, and have a custom connection fetcher and connection validator
The other option I see is to use another annotation

tdraier replied:
Ok for me with the default implementation of getEdges and getPageInfo

tdraier commented on src/main/java/graphql/annotations/connection/simple/SimpleRelay.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public GraphQLObjectType edgeType(String name, GraphQLOutputType nodeType, GraphQLInterfaceType nodeInterface, List<GraphQLFieldDefinition> edgeFields) {
+        return newObject()
```
That is quite confusing, the type here is more the "Connection" type, and edge type is just not used as you don't have edges in this relay implementation (so could be an empty type or null)

guy120494 replied:
Again, I tried to use the existing and modify its behavior. The code calling this method, so I have to return this "type" even though it is not edge

tdraier commented on src/main/java/graphql/annotations/connection/simple/SimplePaginatedDataConnectionFetcher.java:
```diff
@@ -0,0 +1,18 @@
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+
+public class SimplePaginatedDataConnectionFetcher<T> implements SimpleConnectionFetcher<T> {
```
SimpleConnectionFetcher interface is not needed , you directly use SimplePaginatedDataConnectionFetcher in the annotation

tdraier commented on src/main/java/graphql/annotations/connection/simple/SimplePaginatedData.java:
```diff
@@ -0,0 +1,12 @@
+ *
+ * @param <T> the data of which we paginated over
+ */
+public interface SimplePaginatedData<T> extends Iterable<T>, SimpleConnection<T> {
```
Again this one requires getEdges and getPageInfo. The user should only need to return an object with getData() and getTotalCount()

guy120494 replied:
The problem is that any connection fetcher must implement `DataFetcher<Connection<T>>`, and the `Connection` interface has these methods. I can throw an `UnsupportedMethod` exception in the abstract class, or in the interface

tdraier replied:
ok with the UnsupportedMethod exception

tdraier commented on src/main/java/graphql/annotations/processor/GraphQLAnnotations.java:
```diff
@@ -1,12 +1,12 @@
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
  * Unless required by applicable law or agreed to in writing, software
```
Broken licence ?

tdraier reviewed (commented):
Looks great, thanks for the update !
I made some comments inline. Globally I think the "simple" connection defines too many interfaces/abstract/classes which makes it very confusing and hard to follow - simple classes are probably enough most of the time for SimpleFetcher, SimplePaginatedData and SimpleConnection .

guy120494 committed:
Throws the exceptions in the interface

guy120494 committed:
Added annotation to simplify the use of "simpleConnection"

tdraier commented on src/main/java/graphql/annotations/connection/simple/SimpleGraphQLConnection.java:
```diff
@@ -0,0 +1,14 @@
+@Retention(RetentionPolicy.RUNTIME)
+@GraphQLConnection(connectionFetcher = SimplePaginatedDataConnectionFetcher.class, validator = SimplePaginatedDataConnectionTypeValidator.class, connectionType = SimpleRelay.class)
+public @interface SimpleGraphQLConnection {
+}
```
Nice addition, makes things even more simple - can you update the unit test to use this annotation ?

guy120494 replied:
Done

guy120494 committed:
Rename the annotation

guy120494 committed:
Use "GraphQLSimpleConnection"

guy120494 committed:
Deleted "GraphQLSimpleConnection"

guy120494 committed:
Changed the way the union resolver matches an implementation.
Now its by name (if the name of the answer from the datafetcher is the same as one of the candidates)

guy120494 committed:
put back the license

guy120494 committed:
Fix license

guy120494 committed:
Fix license

guy120494 committed:
Undo

guy120494 committed:
Merge branch 'master' into custom-paging

guy120494 committed:
Updated license plugin version
